[2단계 - 장기] 리원(김지윤) 미션 제출합니다

awrion3 opened the pull request:
## 체크 리스트

- [x] 미션의 필수 요구사항을 모두 구현했습니다.
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했습니다.
- [x] 애플리케이션이 정상적으로 실행되었습니다.
- [x] [프롤로그](https://prolog.techcourse.co.kr/studylogs/4122)에 셀프 체크를 작성했습니다.

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.

- 스텝 2에서는 궁성 영역과 점수 계산, DB를 통한 이전 게임 내역 반영 기능을 구현했습니다.
- 스텝 1에서 부족했던 기물 이동 등에 대한 테스트 케이스를 추가로 작성했습니다.
- 하나의 객체가 많은 책임과 역할을 지니지 않도록 분리하고자 시도했습니다.
- 다만 게터를 사용하지 않거나, 인덴트를 1수준으로 유지하는 것은 지키지 못했습니다.

## 어떤 부분에 집중하여 리뷰해야 할까요?

1. 이전 게임 진행 내역 데이터 백업 기능의 추가
이번 스텝 2를 통해 처음으로 도커를 사용해서 mysql 데이터베이스를 실행하고 연동해보게 되었습니다. 그리고 게임 진행 내역을 저장하고 불러오기 위해 BoardStatus 클래스를 도입해서 관련 SQL 메서드들을 구현해보았습니다.

2. 궁성 영역의 구현 관련 팩토리 메서드의 사용
기존에는 추상 클래스와 상속을 사용하여 초 진영과 한 진영의 궁성 영역을 구현하고자 했습니다. 다만 단순히 진영에 따라 구분되는 궁성 영역의 초기화를 위해 불필요하게 상속 구조를 사용한다는 생각이 들게 되었으며, 이를 개선하기 위해 진영별 궁성 영역의 생성을 위한 팩토리 메서드를 도입해보게 되었습니다.

```
안녕하세요 이프! 😄
스텝 2에서는 지난번에 말씀해주신 피드백을 바탕으로 다음과 같이 리팩토링을 진행했습니다:
  1. Manager 클래스의 책임을 GameSetup, GameTracker, GameValidator 클래스로 분산
  2. Palace 클래스 등에서 사용되던 불필요한 추상 클래스 대신 Factory 클래스를 도입
  3. Piece 관련 클래스들에서 사용되던 이동 규칙 validateMovement 메서드의 구조 개선
주말인 와중에 시간 내어 좋은 리뷰 해주셔서 감사합니다! ☺️
```


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- **New Features**
  - 게임 진행 내역 관리 기능을 추가하여 복원, 저장, 삭제, 확인, 생성 기능을 지원합니다.
  - 게임 시작 시 이전 기록 복원 여부를 묻는 새로운 입력 단계가 도입되어 유연한 게임 진행이 가능합니다.
  
- **Documentation**
  - 기능 구현 목록과 입력 예시가 업데이트되어 사용자 안내가 보다 명확해졌습니다.
  
- **Chores**
  - MySQL 데이터베이스 지원 및 도커 환경 설정이 개선되어 시스템 안정성과 데이터 관리가 강화되었습니다.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

awrion3 committed:
refactor: improve piece movement code structure

- add piece movement test case for each piece

awrion3 committed:
refactor: apply palace factory instead of abstract class

awrion3 committed:
refactor: add team name convert method

- add team factory test case, use palace factory class

awrion3 committed:
feat: add board status class for database handling

- add piece factory class and test, remove position range validation

awrion3 committed:
docs: update README.md feature list

awrion3 committed:
refactor: relocate methods in manager to new classes

- use string constant in board status, add line breaks in piece factory

awrion3 committed:
feat: add docker compose setup file

sinb57 commented on src/main/java/janggi/db/BoardStatus.java:
```diff
@@ -0,0 +1,121 @@
+                  );
+                """;
+
+        try (final var conn = getConnection();
+             PreparedStatement psmt = conn.prepareStatement(sql)) {
```
`final` 선언의 기준은 무엇이고 `var` 선언의 기준은 무엇인지 궁금해요~

awrion3 replied:
final은 주로 변수 값의 불변성을 보장하고 싶을 때 사용하고, var는 타입 추론을 통해서 간결하게 선언하고자 할 때 사용하게 되는 것 같아요. 그리고 final var를 적용하면 타입 추론을 활용하면서도 동시에 변수 값을 변경하지 않도록 보장할 수 있게 되는 것 같습니다. 그래서 해당 부분에서 Connection 타입이 try-with-resource 문에서 변경되지 않도록 connection 변수에 대해 final var를 적용해보게 되었습니다.

sinb57 commented on src/main/java/janggi/db/BoardStatus.java:
```diff
@@ -0,0 +1,121 @@
+             PreparedStatement pstmt = conn.prepareStatement(sql)) {
+
+            for (Piece piece : allPieces) {
+                pstmt.setString(1, piece.getName());
+                pstmt.setString(2, piece.getTeamName());
+                pstmt.setString(3, piece.getStatus().name());
```
인덱스가 아니라 파라미터명으로 접근할 수는 없을까요?

awrion3 replied:
음 네, 1씩 증가하는 인덱스 값에 대해 enum을 사용해서 파라미터명으로 접근할 수 있게 만들어보겠습니다!

sinb57 replied:
아 생각해봤는데 java.sql 에 대한 의존이다보니, 제가 생각했던 Named 기반의 PreparedStatement가 제공되지 않더라고요. 그럼에도 enum을 통해 개선해보셔서 감사합니다 🙇‍♂️

enum을 적용해보니 어떻던가요?

awrion3 replied:
음 네, 단순히 인덱스 값을 사용할 때는 숫자 값을 제대로 지정하는 것에 신경을 써야 했는데요. 그 대신 enum을 사용한 경우, 직관적으로 이름을 통해 필요한 값을 설정할 수 있게 되어 좀 더 편의성이 개선된 것 같았습니다!

sinb57 commented on src/main/java/janggi/db/BoardStatus.java:
```diff
@@ -0,0 +1,121 @@
+    public void saveBoardStatus(List<Piece> allPieces) {
+        String sql = "INSERT INTO board_status (piece_name, team_name, piece_status, position_x, position_y) VALUES (?, ?, ?, ?, ?)";
+
+        try (final var conn = getConnection();
+             PreparedStatement pstmt = conn.prepareStatement(sql)) {
```
약어 사용은 지양해주세요~

awrion3 replied:
아 네, 약어 사용하던 부분 수정했습니다!

sinb57 commented on src/main/java/janggi/db/BoardStatus.java:
```diff
@@ -0,0 +1,121 @@
+                pieces.add(PieceFactory.createPiece(pieceName, teamName, pieceStatus, positionX, positionY));
+            }
+        } catch (SQLException e) {
+            e.printStackTrace();
```
`printStackTrace` 사용은 지양해주세요!

awrion3 replied:
앗 네, 해당 부분도 사용하지 않는 방향으로 리팩토링 했습니다!

sinb57 commented on src/main/java/janggi/db/BoardStatus.java:
```diff
@@ -0,0 +1,121 @@
+        }
+    }
+
+    public void createBoardStatus() {
+        String sql = """
+                CREATE TABLE IF NOT EXISTS board_status (
```
docker-compose 의 init을 활용해보면 어떨까요

awrion3 replied:
아 네, 이렇게도 할 수 있었군요. 해당 파일을 활용하는 방식으로 수정했습니다!

sinb57 commented on src/main/java/janggi/game/Manager.java:
```diff
@@ -1,108 +1,47 @@
-
-            Team currentTeam = oldTeam.equals(teamCho) ? teamHan : teamCho;
+            gameTracker.displayBoard(teamHan, teamCho);
+            Team currentTeam = gameTracker.switchTurn(oldTeam, teamHan, teamCho);
             oldTeam = currentTeam;
```
teamHan, teamCho 둘 다 Teams라는 일급 컬렉션에 묶어 두고, switchTurn이나 getCurrentTeam 같은 행위를 부여하면 어떨까요?

awrion3 replied:
아 네, Teams와 같은 일급 컬렉션을 도입하는 것이 더 깔끔 하겠네요, 해당 방향으로 수정해보겠습니다!

sinb57 commented on src/main/java/janggi/game/Manager.java:
```diff
@@ -1,108 +1,47 @@
 
-            Map<String, Position> startingPieceInfo = validateStartingPoint(currentTeam);
+            Map<String, Position> startingPieceInfo = gameValidator.validateStartPoint(currentTeam);
             String pieceName = startingPieceInfo.keySet().iterator().next();
             Position currentPosition = startingPieceInfo.get(pieceName);
```
`keySet().iterator().next()`의 의도가 궁금해요~

awrion3 replied:
아 네, Manager 클래스의 리팩토링 과정에서 해당 부분은 불필요하게 되어 제거되었습니다!

sinb57 commented on src/main/java/janggi/game/GameTracker.java:
```diff
@@ -0,0 +1,50 @@
+import janggi.view.Input;
+import janggi.view.Output;
+
+public class GameTracker {
```
객체의 역할과 책임이 궁금해요

awrion3 replied:
오 네, 해당 GameTracker 클래스의 경우 여러 책임을 가지고 있다고 생각이 되네요. 해당 클래스에 존재하던 메서드들은 Teams, GameProcessManager 클래스 등으로 분산되어 이동하게 되었습니다!

sinb57 commented on src/main/java/janggi/game/Manager.java:
```diff
@@ -1,108 +1,47 @@
             Position currentPosition = startingPieceInfo.get(pieceName);
-
-            Position destination = validateDestination(boardNavigator, currentTeam, pieceName, currentPosition);
+            Position destination = gameValidator.validateDestination(currentTeam, pieceName, currentPosition);
```
`validate`와 달리, 사용자로부터 입력값을 받아 반환하네요. 검증한다는 이름과 달리 입력을 받아 반환하는 행위가 섞여 있어서 의도 파악이 어려웠어요. 내부 구현을 추측할 수 있도록 적절한 네이밍을 고려해보세요.

awrion3 replied:
음 네, 입력 값에 대한 검증 외에도 사용자로부터 입력을 요청한다는 의미도 메서드명에 반영될 수 있도록 수정해보겠습니다!

sinb57 commented on src/main/java/janggi/palace/PalaceFactory.java:
```diff
@@ -0,0 +1,34 @@
+
+public class PalaceFactory {
+    public static Palace createPalace(TeamName teamName) {
+        if (teamName.matchTeamName("초")) {
+            return new Palace(List.of(
+                    new Position(3, 0),
```
else 가 안쓰였을 뿐, 사실상 if-else 네요. Enum의 다형성으로 풀어보세요.

sinb57 replied:
더불어 TeamScore#calculateTeamScore 또한 같이 고민해보시고,
if-else의 사용을 제한하는 객체 지향 생활 체조는 무엇을 학습하길 의도한 것인지 고민해보셨으면 좋겠어요~

awrion3 replied:
오 네, 객체 지향을 위해 enum으로도 다형성을 적용해볼 수 있었군요. PalaceFactory와 TeamScore 클래스에서 관련 부분 수정해보도록 하겠습니다!

sinb57 commented on src/main/java/janggi/piece/Cannon.java:
```diff
@@ -10,26 +10,45 @@ public Cannon(TeamName teamName, Position position) {
-        if (palaceArea == PalaceArea.OUTSIDE) {
-            isValidMove = offsetX == OFFSET_ZERO && offsetY <= OFFSET_Y_MAX
-                    || offsetX <= OFFSET_X_MAX && offsetY == OFFSET_ZERO;
+        if (palaceArea.equals(PalaceArea.OUTSIDE) && isValidMoveOutsidePalace(offsetX, offsetY)) {
```
`palaceArea.equals(PalaceArea.OUTSIDE)` 이 로직은 PalaceArea가 수행토록 해도 좋겠네요. isOutside / isInside 처럼요

awrion3 replied:
아 네, PalaceArea 클래스에 isOutside(), isInside() 메서드를 추가해서 사용하는 방식으로 리팩토링 해보겠습니다!

sinb57 reviewed (changes requested):
안녕하세요 리원~ 2단계 진행 하느라 고생 많으셨어요!
기존의 Manager를 잘 세분화 해보셨네요👍  코멘트 확인해보시고 반영해주세요~
그리고 미션 요구사항 중 `객체지향 생활 체조 원칙`를 지키도록 리팩터링 마저 해주세요

awrion3 committed:
refactor: use init sql file to create table

- use board status column enum for index values

awrion3 committed:
refactor: add and use palace area boolean methods

- use piece name enum for string piece name

awrion3 committed:
refactor: use enum polymorphism for factory class

awrion3 committed:
refactor: use teams class instead of team list

- rename request and validation methods, relocate check turn and opponent methods

sinb57 commented on src/main/java/janggi/game/GameController.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    public void run() {
+        Teams teams = gameProcessManager.initializeTeams();
+        Team teamCho = teams.getTeamCho();
+        Team teamHan = teams.getTeamHan();
+        Team oldTeam = teamHan;
+        do {
+            output.printBoard(teamHan, teamCho);
+            Team currentTeam = teams.switchTurn(oldTeam);
+            oldTeam = currentTeam;
+
+            Piece currentPiece = gameRequestValidator.requestAndValidateStartPoint(currentTeam);
+            Position destination = gameRequestValidator.requestAndValidateDestination(currentTeam, currentPiece);
+            String pieceName = currentPiece.getName();
+            Position currentPosition = currentPiece.getPosition();
+
+            currentTeam.move(pieceName, currentPosition, destination);
+            Team opponentTeam = teams.checkOpponent(currentTeam);
+            opponentTeam.updateStatusIfCaught(destination);
+        } while (gameProcessManager.isContinue(teamHan, teamCho));
+
+        output.printTeamScore(teamCho.checkTeamScore(), teamHan.checkTeamScore());
+        gameProcessManager.saveTeams(teamCho, teamHan);
```
https://github.com/woowacourse/java-janggi/pull/113#discussion_r2020189148 코멘트에서 아래와 같은 변경을 기대했답니다.

```suggestion
        Teams teams = gameProcessManager.initializeTeams();
        do {
            output.printBoard(teams);
            Team currentTeam = teams.getCurrentTeam();

            Piece currentPiece = gameRequestValidator.requestAndValidateStartPoint(currentTeam);
            Position destination = gameRequestValidator.requestAndValidateDestination(currentTeam, currentPiece);
            String pieceName = currentPiece.getName();
            Position currentPosition = currentPiece.getPosition();

            currentTeam.move(pieceName, currentPosition, destination);
            Team opponentTeam = teams.checkOpponent(currentTeam);
            opponentTeam.updateStatusIfCaught(destination);

            teams.switchTurn();
        } while (gameProcessManager.isContinue(teams));

        output.printTeamScore(teams);
        gameProcessManager.saveTeams(teams);
```

awrion3 replied:
오 네, `Teams` 클래스를 좀 더 활용하는 방식으로 `GameController` 클래스의 구조를 개선 해볼 수 있었네요. 해당 부분을 반영해서 좀 더 리팩토링 해보겠습니다!

sinb57 commented on src/main/java/janggi/game/GameController.java:
```diff
@@ -0,0 +1,46 @@
+
+            Piece currentPiece = gameRequestValidator.requestAndValidateStartPoint(currentTeam);
+            Position destination = gameRequestValidator.requestAndValidateDestination(currentTeam, currentPiece);
+            String pieceName = currentPiece.getName();
+            Position currentPosition = currentPiece.getPosition();
+
+            currentTeam.move(pieceName, currentPosition, destination);
```
`currentTeam.move(`에 currentPiece.getName()가 아니라 currentPiece를 넘기면 안되나요??

awrion3 replied:
앗 네, currentPiece를 넘기는 방식을 적용할 수 있었네요. 해당 방식으로 수정했습니다!

sinb57 commented on src/main/java/janggi/game/GameProcessManager.java:
```diff
@@ -0,0 +1,83 @@
+
+    private Teams loadTeamsFromBoardStatus() {
+        List<Piece> allPieces = boardStatus.loadBoardStatus();
+        List<Piece> choPieces = allPieces.stream().filter(piece -> piece.getTeamName().equals("초")).toList();
```
문자열을 비교하는 것은 자칫 오타 등으로 인해 버그가 발생할 수 있어요. TeamName Enum 값을 활용해보세요

awrion3 replied:
오 네, 기존에 존재하던 Piece의 `matchTeam()` 메서드를 사용해 TeamName 값을 비교하는 방식으로 다시 수정했습니다!

sinb57 commented on src/main/java/janggi/game/GameRequestValidator.java:
```diff
@@ -0,0 +1,53 @@
+        while (true) {
+            try {
+                Map<String, Position> pieceStartingPoint = input.readPieceStartPoint(currentTeam);
+                String pieceName = pieceStartingPoint.keySet().iterator().next();
```
https://github.com/woowacourse/java-janggi/pull/113#discussion_r2020189601
`keySet().iterator().next()`의 의도가 궁금해요~

awrion3 replied:
아 네, 해당 코드 부분은 맵 자료구조에서 키 값을 추출하는 과정에서 사용하게 되었어요. pieceStartingPoint의 경우, 항상 단 하나의 기물 이름(키)와 기물 위치(값)으로만 이루어져서 유일한 키를 가져오기 위해 해당 방식을 적용해보게 되었습니다.

sinb57 replied:
`Input#readPieceStartPoint` 반환 타입이 Map 이긴 하지만, 요소가 하나 뿐인 Map이네요. pieceName과 pieceStartPoint을 함께 반환하기 위해 Map으로 감싼 것 같아요. 이럴 때는 DTO를 만들어서 전달하는 것이 훨씬 좋아요.

sinb57 commented on src/main/java/janggi/game/GameRequestValidator.java:
```diff
@@ -0,0 +1,53 @@
+        this.input = input;
+    }
+
+    public Piece requestAndValidateStartPoint(Team currentTeam) {
```
`requestValidatedStartPoint`와 같은 네이밍도 괜찮아보여요

awrion3 replied:
오 네, 메서드명에 and를 넣은 기존 네이밍에 비해 좀 더 간단해 졌네요, 해당 네이밍 방식으로 변경하도록 하겠습니다!

sinb57 commented on src/main/java/janggi/db/BoardStatus.java:
```diff
@@ -0,0 +1,102 @@
+        try {
+            return DriverManager.getConnection("jdbc:mysql://" + SERVER + "/" + DATABASE + OPTION, USERNAME, PASSWORD);
+        } catch (final SQLException e) {
+            System.out.println(INVALID_DB_CONNECTION + e.getMessage());
```
error 문구를 `System.out.println` 출력하는 행위는 지양해주세요~ 가급적 Exception으로 던지는 것이 옳아요

awrion3 replied:
아 네, 런타임 예외를 던지는 방식으로 변경해보도록 하겠습니다!

sinb57 commented on src/main/java/janggi/palace/PalaceFactory.java:
```diff
@@ -0,0 +1,49 @@
+import java.util.List;
+
+public enum PalaceFactory {
+    TEAM_CHO {
+        @Override
+        public Palace createTeamPalace(TeamName teamName) {
+            return new Palace(List.of(
+                    new Position(3, 0),
+                    new Position(4, 0),
+                    new Position(5, 0),
+                    new Position(3, 1),
+                    new Position(4, 1),
+                    new Position(5, 1),
+                    new Position(3, 2),
+                    new Position(4, 2),
+                    new Position(5, 2)
+            ));
+        }
+    },
```
메서드를 override 하는 방식보단, Palace를 생성해서 지니고 있으면 어떨까요? 마침 Palace가 불변 객체네요

awrion3 replied:
아 네, 굳이 메서드를 통한 생성 방식보다는 Palace를 미리 생성해서 갖고 있는 방식이 더 깔끔 하겠네요, 해당 부분 리팩토링 해보겠습니다!

sinb57 reviewed (changes requested):
안녕하세요 리원~ 리뷰 반영하느라 고생 많으셨어요!
추가 코멘트 확인 부탁드려요~

awrion3 committed:
refactor: use palace constructor in palace factory

- use exception in creating jdbc connection, rename methods in game request validator

awrion3 committed:
refactor: use teams parameter in game controller

- use enum team name in game process manager, use piece as parameter value for move method in team, use constants in output

sinb57 reviewed (approved):
안녕하세요 리원~ 장기 미션 진행하느라 고생 많으셨어요!!
아쉽지만 리뷰 요청 마감 기한이 지났기 때문에 리뷰어 가이드에 따라 PR 닫을게요
계속 얘기나누고 싶은 리뷰 내용이 있거나 질문이 있다면, 언제든 편하게 DM 주세요~
